File: rymobile/frameworks/runtime-src/Classes/tools/manager/Game_Path_Manager.cpp
//
//  Game_Path_Manager.cpp
//  qipaicocos
//
//  Created by keven on 16/8/12.
//
//
#include <stdio.h>
#include "Game_Path_Manager.h"
#include "cocos2d.h"
#include "../DictionaryAide.h"

USING_NS_CC;
using namespace std;
//////////////////////////////////////////////////////////////////////////

Game_Path_Manager* Game_Path_Manager::msInstance = 0;

Game_Path_Manager* Game_Path_Manager::shared()
{
    if (msInstance == 0)
    {
        msInstance = new Game_Path_Manager();
    }
    
    return msInstance;
}

void Game_Path_Manager::purge()
{
    if (msInstance)
        delete msInstance;
    msInstance = 0;
}

///////////////////////////////////////////////////////////////////////////
Game_Path_Manager::Game_Path_Manager()
{
}

Game_Path_Manager::~Game_Path_Manager()
{
	map<int,vector<Move_Points>>::iterator it = _map.begin();
	if(it != _map.end()){
		it->second.clear();
	}
}

size_t game_pfwrite(void* data, int elemsize, int count, std::string &sz_path)
{
    unsigned char dataa[1024];
    int size_byte = elemsize*count;
    
    unsigned char* src = (unsigned char*)data;
    unsigned char* dest = dataa;
    
    for (int i = 0; i < size_byte; i++)
        *(dest++) = (*src++) ^ 0xDE;
    
    dataa[size_byte] = '\0';
    sz_path += (char*)dataa;
    
    return size_byte;
}

bool game_get_string_line(std::string& path, const std::string& all, unsigned int& start_pos)
{
    path.clear();
    int end_pos = 0;
    unsigned int  n = all.length();
    
    if ((start_pos >= 0) && (start_pos < n))
    {
        end_pos = all.find_first_of("\n ", start_pos);
        
        if ((end_pos < 0) && (end_pos >n))
            end_pos = n;
        
        path += all.substr(start_pos, end_pos - start_pos);
        start_pos = end_pos + 1;
        
    }
    return !path.empty();
}

bool game_path_load(std::vector<Move_Points>& paths, int count, const char* format)
{
    paths.clear();
    std::string sData;
    std::string line;
    
    for (int i = 0; i < count; i++)
    {
        
        char filename[128] = { 0 };
        sprintf(filename, format, i);
        //获得文件在系统的绝对路径
        std::string filepath = CCFileUtils::sharedFileUtils()->fullPathForFilename(filename);
        //读取的字节数，读取失败则为0
        ssize_t nread = 0;
        unsigned char * rdata = CCFileUtils::sharedFileUtils()->getFileData(filepath.c_str(), "rb", &nread);
        if (nread <= 0 || rdata == 0)
            return false;
        
        //解密
        sData.clear();
        sData.resize(nread, 0);
        for (int i = 0; i < nread; i++)
        {
            //printf(" %c" , rdata[i]);
            sData[i] = rdata[i] ^ 0xDE;
        }
        delete[] rdata;
    
        
        
        //新加路径
        paths.push_back(Move_Points());
        Move_Points &move_points = paths[paths.size() - 1];
        
        //每行读取
        unsigned int start_pos = 0;
        game_get_string_line(line, sData, start_pos);
        
        while (game_get_string_line(line, sData, start_pos))
        {
            int x, y, staff;
            float angle;
            //读取路径节点信息
            sscanf(line.c_str(), "(%d,%d,%f,%d)", &x, &y, &angle, &staff);
            //CCLOG("(%d,%d,%d,%d)", i, x, y, staff);
            move_points.push_back(Move_Point(xPoint(x, y), angle));
        }
    

    }
    
    return true;
}

static __Array* visitArray(const ValueVector& array);
static __Dictionary* visitDict(const ValueMap& dict)
{
	__Dictionary* ret = new __Dictionary();
	ret->init();

	for (auto iter = dict.begin(); iter != dict.end(); ++iter)
	{
		if (iter->second.getType() == Value::Type::MAP)
		{
			const ValueMap& subDict = iter->second.asValueMap();
			auto sub = visitDict(subDict);
			ret->setObject(sub, iter->first);
			sub->release();
		}
		else if (iter->second.getType() == Value::Type::VECTOR)
		{
			const ValueVector& arr = iter->second.asValueVector();
			auto sub = visitArray(arr);
			ret->setObject(sub, iter->first);
			sub->release();
		}
		else
		{
			auto str = new __String(iter->second.asString());
			ret->setObject(str, iter->first);
			str->release();
		}
	}
	return ret;
}

static __Array* visitArray(const ValueVector& array)
{
	__Array* ret = new __Array();
	ret->init();

	for (const auto &value : array) {
		if (value.getType() == Value::Type::MAP)
		{
			const ValueMap& subDict = value.asValueMap();
			auto sub = visitDict(subDict);
			ret->addObject(sub);
			sub->release();
		}
		else if (value.getType() == Value::Type::VECTOR)
		{
			const ValueVector& arr = value.asValueVector();
			auto sub = visitArray(arr);
			ret->addObject(sub);
			sub->release();
		}
		else
		{
			auto str = new __String(value.asString());
			ret->addObject(str);
			str->release();
		}
	}

	return ret;
}

bool Game_Path_Manager::initialise_paths(const std::string &directory)
{
	std::string fullPath = FileUtils::getInstance()->fullPathForFilename(directory);
	ValueVector arr2_ = FileUtils::getInstance()->getValueVectorFromFile(fullPath);

	__Array *arr2 = visitArray(arr2_);
	if (arr2){
		arr2->autorelease();
	}

	for (int j = 0, jl = (int)arr2->count(); j < jl; ++j)
	{
		__Dictionary* ani = (__Dictionary*)arr2->getObjectAtIndex(j);
		if(ani){
			DictionaryAide aide(ani);

			int type			= aide.intFromKey("path_type");
			int num				= aide.intFromKey("path_num");	
			const char* path	= aide.stringFromKey("path");

			if(_map.find(type) == _map.end()){

				vector<Move_Points>   save_paths_;
				game_path_load(save_paths_,num,path);

				_map.insert(pair<int,vector<Move_Points>>(type,save_paths_));
			}
		}
	}

    return true;
}



Move_Points &Game_Path_Manager::get_paths(unsigned short path_id, unsigned char path_type)
{
	 map<int,vector<Move_Points>>::iterator it = _map.find(path_type);
	if(it != _map.end()){

		vector<Move_Points> & vecPoints = it->second;
		if(vecPoints.size() > path_id){
			return vecPoints[path_id];
		}
	}

	return movePointsNull;	
}


///////////////////////////////////////////////////////////////////////////

File: rymobile/client/game/yule/lkpy/src/models/Game_Path_Manager.lua
--
-- Author: Liuxu
-- Date: 2018-08-25 10:26:25
-- 路径管理
local module_pre = "game.yule.lkpy.src"	

local Game_Path_Manager = class("Game_Path_Manager")
local Move_Poiots 		= appdf.req(module_pre .. ".models.Move_Point_Manager")
local Action_Move		= appdf.req(module_pre .. ".models.Action_Move_Point")

appdf.req(module_pre .. ".models.god")
appdf.req(module_pre .. ".models.god2")


msInstance = nil

function Game_Path_Manager:shared()
    
    if nil == msInstance then
		msInstance = self
		self._paths  = nil
		self._map   = {}
    end
    return msInstance
end

function Game_Path_Manager:ctor()
    print("初始化")
    return
end

function Game_Path_Manager:game_get_string_line( path , all , start_pos )
	path = ""
	local end_pos = 0
	local n = string.len( all )
	if start_pos >= 0 and start_pos < n then
		end_pos = string.find(all, "\n" , start_pos)
		if end_pos < 0 and end_pos > n then
			end_pos = n
		end

		local a =  string.sub( all, start_pos,end_pos - start_pos )
		path = path .. a
		
	end

	--self.start_pos = end_pos +1
	self._sData = string.sub( all,end_pos+1 ,-1)
	self.line = path 
	if path == "" then
		return false
	else
		return true
	end
	
end


function Game_Path_Manager:game_path_load( paths ,count , format )

	
	paths = {}
	local sData	= {}
	self.line = ""

	for i=0,count do
		local str_ = string.format(format,i)
		--获取绝对路径
		local filepath = cc.FileUtils:getInstance():fullPathForFilename(str_)
		local sData		= cc.FileUtils:getInstance():getStringFromFile(filepath)

		function split( str,reps )
			local resultStrList = {}
			string.gsub(str,'[^'..reps..']+',function ( w )
				table.insert(resultStrList,w)
			end)
			return resultStrList
		end

		--每行读取
		self.start_pos = 0;
		self._sData = sData
		--dump(sData)
		self:game_get_string_line(self.line,self._sData,self.start_pos)

		local paths2 = {}

		while self:game_get_string_line(self.line,self._sData,self.start_pos) do
			local x,y,staff , angle 
			--dump(self.line)
			--dump(self.start_pos)
			--print(type(self.line))
			local indexfirst  = 1
			local indexend = string.find(self.line,")")
			if indexend == nil then
				dump(self.line)
				dump(self._sData)
			end
			local str = string.sub(self.line,indexfirst+1,indexend-1)
			--print("" .. str)
			local tab_str = split(str,",")
			x = tab_str[1]
			y = tab_str[2]
			angle = tab_str[3]
			staff = tab_str[4]
			
			local move_points  = Move_Poiots:create(x,y,angle)
			--dump(move_points)
			table.insert( paths2, move_points)
		end
		--dump(paths)
		table.insert( paths, paths2)
	end

	self._paths = paths
	
    return true;
end

function Game_Path_Manager:initialise_paths( )
	local directory = "resources_path.plist"

	local fullpath =  cc.FileUtils:getInstance():fullPathForFilename(directory)
	
	local arr4_	   = cc.FileUtils:getInstance():getValueVectorFromFile(fullpath)
	dump(arr4_)

	for i=1,#arr4_ do
		if arr4_[i] ~= nil then
			local type = arr4_[i]["path_type"]
			local num  = arr4_[i]["path_num"]
			local path = arr4_[i]["path"]
			--if self._map[type] == nil then
				local save_paths = {}
				local save_paths2 
				--self:game_path_load(save_paths,num,path)
				if i == 1 then
					save_paths2 = god["1"]
				elseif i == 2 then 
					 save_paths2 = god2["2"]
				elseif i == 3 then 
					 save_paths2 = god2["3"]
				elseif i == 4 then 
					 save_paths2 = god2["4"]
				elseif i == 5 then 
					 save_paths2 = god2["5"]	
				end
				
				--print(save_paths2)
				--local save_path3 = {}
				--save_path3[type] = save_paths2
				--dump(save_path3)
				self._map[type] = save_paths2
				--table.insert( self._map, save_path3)
			--end
		end
	end


    return true;
end


function Game_Path_Manager:get_paths( path_id , path_type )

	path_type = "" .. path_type
	local it = self._map[path_type]
	--dump(it)
	for i=1,#it do
		if i > path_id then
			return it[i]
		end
	end
end

return Game_Path_Manager